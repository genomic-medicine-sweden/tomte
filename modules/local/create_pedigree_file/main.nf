process CREATE_PEDIGREE_FILE {
    tag "pedigree"
    label 'process_single'

    conda "conda-forge::python=3.8.3"
    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
        'https://depot.galaxyproject.org/singularity/python:3.8.3' :
        'biocontainers/python:3.8.3' }"

    input:
    val(samples)

    output:
    path("*.ped"),       emit: ped
    path "versions.yml", emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def case_name = samples[0].case
    def outfile_text = ['#family_id', 'sample_id', 'father', 'mother', 'sex', 'phenotype'].join('\\t')
    def samples_list = []
    samples.each { sample ->
        def sample_name = sample.sample
        if (!samples_list.contains(sample_name)) {
            def sex_int = sample.sex in ["M", 1] ? "1" : sample.sex in ["F", 2] ? "2" : "0"
            // If paternal or maternal ID is null, i.e. not in dataset, set it to "0"
            def paternal = sample.paternal ?: "0"
            def maternal = sample.maternal ?: "0"
            outfile_text += "\\n" + [sample.case, sample_name, paternal, maternal, sex_int, sample.phenotype].join('\\t')
            samples_list.add(sample_name)
        }
    }
    """
    echo -e "$outfile_text" > ${case_name}.ped

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        create_pedigree_file: v1.1
        python: \$(python --version | sed 's/Python //g')
    END_VERSIONS
    """

    stub:
    def case_name = samples[0].case
    """
    touch ${case_name}.ped

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        create_pedigree_file: v1.1
        python: \$(python --version | sed 's/Python //g')
    END_VERSIONS
    """
}
