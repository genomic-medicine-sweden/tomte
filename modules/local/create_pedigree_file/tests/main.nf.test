nextflow_process {

    name "Test Process CREATE_PEDIGREE_FILE"
    script "../main.nf"
    process "CREATE_PEDIGREE_FILE"
    tag "modules_local"
    tag "modules"
    tag "create_pedigree_file"

    test("Trio") {

        when {
            process {
                """
                input[0] =  [
                    [case:'finequagga', id:'ACC5963A1', sample:'ACC5963A1', strandedness:'reverse', paternal:'ACC5963A2', maternal:'ACC5963A3', sex:'M', single_end:false, fq_pairs:1, is_fastq:false],
                    [case:'finequagga', id:'ACC5963A2', sample:'ACC5963A2', strandedness:'reverse', paternal:'0', maternal:'0', sex:'M', single_end:false, fq_pairs:1, is_fastq:true],
                    [case:'finequagga', id:'ACC5963A3', sample:'ACC5963A3', strandedness:'reverse', paternal:'0', maternal:'0', sex:'F', single_end:false, fq_pairs:1, is_fastq:false]
                ]
                """
            }
        }

        then {
            assert process.success
            assert snapshot(process.out).match()
        }

    }

    test("Single sample") {

        when {
            process {
                """
                input[0] =  [
                    [case:'finequagga', id:'ACC5963A1', sample:'ACC5963A1', strandedness:'reverse', paternal:'ACC5963A2', maternal:'ACC5963A3', sex:'M', single_end:false, fq_pairs:1, is_fastq:false],
                ]
                """
            }
        }

        then {
            assert process.success
            assert snapshot(process.out).match()
        }

    }

    test("No parents or sex") {

        when {
            process {
                """
                input[0] =  [
                    [case:'finequagga', id:'ACC5963A1', sample:'ACC5963A1', strandedness:'reverse', paternal:'0', maternal:'0', sex:'NA', single_end:false, fq_pairs:1, is_fastq:false],
                    [case:'finequagga', id:'ACC5963A2', sample:'ACC5963A2', strandedness:'reverse', paternal:'0', maternal:'0', sex:'NA', single_end:false, fq_pairs:1, is_fastq:true],
                    [case:'finequagga', id:'ACC5963A3', sample:'ACC5963A3', strandedness:'reverse', paternal:'0', maternal:'0', sex:'NA', single_end:false, fq_pairs:1, is_fastq:false]
                ]
                """
            }
        }

        then {
            assert process.success
            assert snapshot(process.out).match()
        }

    }

}
