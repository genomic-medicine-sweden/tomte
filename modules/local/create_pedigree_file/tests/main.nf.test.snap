{
    "Trio": {
        "content": [
            {
                "0": [
                    "finequagga.ped:md5,bd5029b95f86b7e64aacb9b744285207"
                ],
                "1": [
                    "versions.yml:md5,7b861aa9c6736cf73ca9f7f18c3521de"
                ],
                "ped": [
                    "finequagga.ped:md5,bd5029b95f86b7e64aacb9b744285207"
                ],
                "versions": [
                    "versions.yml:md5,7b861aa9c6736cf73ca9f7f18c3521de"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.5"
        },
        "timestamp": "2025-04-15T12:01:08.110576"
    },
    "No parents or sex": {
        "content": [
            {
                "0": [
                    "finequagga.ped:md5,d67d056eb3732340d84a8957865ef1ca"
                ],
                "1": [
                    "versions.yml:md5,7b861aa9c6736cf73ca9f7f18c3521de"
                ],
                "ped": [
                    "finequagga.ped:md5,d67d056eb3732340d84a8957865ef1ca"
                ],
                "versions": [
                    "versions.yml:md5,7b861aa9c6736cf73ca9f7f18c3521de"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.5"
        },
        "timestamp": "2025-04-15T12:03:23.02631"
    },
    "Single sample": {
        "content": [
            {
                "0": [
                    "finequagga.ped:md5,c80b46d891408d585e79b1b8389f50fc"
                ],
                "1": [
                    "versions.yml:md5,7b861aa9c6736cf73ca9f7f18c3521de"
                ],
                "ped": [
                    "finequagga.ped:md5,c80b46d891408d585e79b1b8389f50fc"
                ],
                "versions": [
                    "versions.yml:md5,7b861aa9c6736cf73ca9f7f18c3521de"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.5"
        },
        "timestamp": "2025-04-15T12:01:16.408464"
    }
}